### ISSUE TITLE
Disable dsf when piping to file?

### ISSUE THREAD

**Issue by woodrowbarlow** [so-fancy/diff-so-fancy#181]
I have used the suggested git pager replacement from the readme (`git config --global core.pager "diff-so-fancy | less --tabs=4 -RFX"`) so that commands like "git diff" and "git show" use diff-so-fancy by default. Is is possible to write this config option such that diff-so-fancy is only used when the output stream is stdout?

Basically, I want to be able to do `git diff HEAD^..HEAD > changes.patch` and end up with a usable patch file, whereas currently I need to do `git --no-pager diff --no-color HEAD^..HEAD > changes.patch`.

**Comment by stevemao** [comment]
What about alias `git --no-pager`?

**Comment by woodrowbarlow** [comment]
That's what I do now, I was just wondering if anyone more clever than I has managed to work some shell magic that can redirect the pager by default depending on what type of stream is being used.

**Comment by OJFord** [comment]
This is not possible - there's no way d-s-f (or any program) can detect where its output is going to go _before it has produced any output_, I can't imagine how it would even detect where it went _after_ it had produced it.

> such that diff-so-fancy is only used when the output stream is stdout?

If you redirect stdout to a file, it was still stdout. The program is in control of what it's output stream is when it prints something, i.e. usually stdout; stderr for an error message.

As the user you are in control of _redirecting_ those streams, but you can't change what they were to begin with.


If you're using both workflows frequently, I'd suggest adding an alias to your `.gitconfig`:
```ini
[alias]
    patch = --no-pager diff --no-color
```

So that you can do `git diff` for d-s-f goodness, and `git patch > changes.patch` for standard rawness.

**Comment by scottchiefbaker** [comment]
In Perl I know you can use `-t` to check if the output is interactive (i.e. not piped to a file).

**Comment by OJFord** [comment]
My apologies, @scottchiefbaker is right, `-t` is not just Perl but POSIX.

Maybe we can and should do this?

**Comment by OJFord** [comment]
@woodrowbarlow I've had a crack at this, and I _think_ it's fully functional.

Please check it out and let me know if it behaves as you expect :)
(Only change is to the bash script in the root dir, so you can just grab that; stick it in a git repo, and run `git diff --no-pager | ./diff-so-fancy [> changes.patch]`.)

**Comment by scottchiefbaker** [comment]
Is this fixed?

**Comment by OJFord** [comment]
Haven't touched this for a while, implementation is in #190, I believe it's working but needs some modifications to existing tests (as well as tests of the new functionality). I'll check where it's up to and have a look at tests over the weekend.

**Comment by scottchiefbaker** [comment]
It works for me... If I do 

```
git diff > /tmp/out.patch
```

I get a raw diff patch with nothing fancy added. In other words "works for me", so I'd be inclined to close this issue. Let me know when you're ready.

**Comment by OJFord** [comment]
Works, with #190, do you mean? Surely without you get colour codes, etc.

**Comment by scottchiefbaker** [comment]
I'm using `master` and I don't have this issue.

**Comment by OJFord** [comment]
Looks like there's been several PRs merged and other commits since the last release. [0]

That might explain some of the issues being opened for things already fixed, if their openers like myself ordinarily track the released version.

Sounds like this was fixed/implemented in parallel to my #190 then, so that can also be closed. I don't think it was any wider in scope than what's described here, but I can always reopen it if that turns out to be the case.

[0] - https://github.com/so-fancy/diff-so-fancy/compare/v0.11.2...master

**Comment by scottchiefbaker** [comment]
I'm 99% sure this is addressed. If not feel free to re-open.

#wfm
